### PR TITLE
*: add health handler for grpcproxy self

### PR DESCRIPTION
**Pull request by tangcong**
When grpcproxy hangs or is overloaded, the watch client will not receive any change event  and cannot detect the proxy failure. We need a health check to remove the abnormal proxy in time.

### CODE REVIEW

File: proxy/grpcproxy/health.go [hunk 1]
```diff
@@ -43,3 +51,25 @@ func checkHealth(c *clientv3.Client) etcdhttp.Health {
...
 }
+
+func checkProxyHealth(c *clientv3.Client) etcdhttp.Health {
+	h := etcdhttp.Health{Health: "true"}
+	ctx, cancel := context.WithTimeout(c.Ctx(), time.Second)
+	_, err := c.Get(ctx, "a")
+	cancel()
+	if err != nil && err != rpctypes.ErrPermissionDenied {
+		h.Health = "false"
+		h.Reason = "GET ERROR"
```
Comment (gyuho) [inline proxy/grpcproxy/health.go:62]: Should we include original error message?

File: proxy/grpcproxy/health.go [hunk 2]
```diff
@@ -43,3 +51,25 @@ func checkHealth(c *clientv3.Client) etcdhttp.Health {
 	}
 	return h
 }
+
+func checkProxyHealth(c *clientv3.Client) etcdhttp.Health {
+	h := etcdhttp.Health{Health: "true"}
+	ctx, cancel := context.WithTimeout(c.Ctx(), time.Second)
```
Comment (gyuho) [inline proxy/grpcproxy/health.go:57]: Can we embed `checkHealth` and add extra watch checks?

File: proxy/grpcproxy/health.go (hunk 1 above)
Comment (tangcong) [reply proxy/grpcproxy/health.go]: updated.thanks

File: proxy/grpcproxy/health.go (hunk 2 above)
Comment (tangcong) [reply proxy/grpcproxy/health.go]: updated.thanks.

**Comment by tangcong** [comment]
grpcproxy may have a different certificate than etcd, now `proxy/health` works well when the user enables auto-tls, but i am not sure whether we should add the specified client certificate flags for the health check. proxy cert(grpcProxyListenCert) will be used as a client certificate in health check scenarios. the user-defined proxy certificate may cause the health check to time out.  @gyuho

**Comment by gyuho** [comment]
> proxy cert(grpcProxyListenCert) will be used as a client certificate in health check scenarios

Think this is good for now.

**Comment by tangcong** [comment]
> > proxy cert(grpcProxyListenCert) will be used as a client certificate in health check scenarios
> 
> Think this is good for now.

ok, thanks. I will follow the feedback from the community. If anyone meets the problem, one of the solutions is that cert usages can include `client auth`.
```
            "server": {
                "expiry": "2540400h",
                "usages": [
                    "signing",
                    "key encipherment",
                    "server auth",
		    "client auth"
                ]
            },
```

**Pull request merged**